What's new in AOCL 3.1

amd-toolchain-support opened the pull request:
1) AMD BLIS:
    1.a) Supports Dynamic Dispatch and AOCL Dynamic feature
    1.b) Improvements in DGEMM, ZGEMM, DTRSM, DSYRK, xGEMV, and DOTV
2) AMD libFLAME:
    2.a) Supports LAPACK 3.10.0 specification
    2.b) Optimized factorization and ZGEEV routines
3) AMD FFTW:
    3.a) Features like 'AMD application optimization layer', 'Fast MPI transpose algorithm' and 'Top N planner' are added
4) AMD LibM:
    4.a) Optimized exp2, log2 (Single and Double precision) scalar and vector
    4.b) Optimized log10f (scalar and vector) and powf vector variants to support WRF4.1.2 benchmark
5) AOCL-Sparse:
    5.a) New API for sparse matrix and dense matrix multiplication
6) AMD ScaLAPACK:
    6.a) ILP64 support has been enabled
7) AOCL enabled MUMPS library:
    7.a) CMake based build system on Windows for AOCL enabled MUMPS sparse solver library will be available shortly on GitHub
        7.a.i)	Refer https://github.com/amd/mumps-build
    7.b) Spack-based recipe on Linux for AOCL enabled MUMPS sparse solver library will be enabled shortly

amd-toolchain-support committed:
What's new in AOCL 3.1

1) AMD BLIS:
    1.a) Supports Dynamic Dispatch and AOCL Dynamic feature
    1.b) Improvements in DGEMM, ZGEMM, DTRSM, DSYRK, xGEMV, and DOTV
2) AMD libFLAME:
    2.a) Supports LAPACK 3.10.0 specification
    2.b) Optimized factorization and ZGEEV routines
3) AMD FFTW:
    3.a) Features like 'AMD application optimization layer', 'Fast MPI transpose algorithm' and 'Top N planner' are added
4) AMD LibM:
    4.a) Optimized exp2, log2 (Single and Double precision) scalar and vector
    4.b) Optimized log10f (scalar and vector) and powf vector variants to support WRF4.1.2 benchmark
5) AOCL-Sparse:
    5.a) New API for sparse matrix and dense matrix multiplication
6) AMD ScaLAPACK:
    6.a) ILP64 support has been enabled
7) AOCL enabled MUMPS library:
    7.a) CMake based build system on Windows for AOCL enabled MUMPS sparse solver library will be available shortly on GitHub
        7.a.i)	Refer https://github.com/amd/mumps-build
    7.b) Spack-based recipe on Linux for AOCL enabled MUMPS sparse solver library will be enabled shortly

amd-toolchain-support committed:
Fix invalid version range error

dev-zero commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -26,41 +26,151 @@ class Amdfftw(FftwBase):
     _name = 'amdfftw'
     homepage = "https://developer.amd.com/amd-aocl/fftw/"
-    url = "https://github.com/amd/amd-fftw/archive/3.0.tar.gz"
+    url = "https://github.com/amd/amd-fftw/archive/3.1.tar.gz"
```
unnecessary

dev-zero replied:
... same for the other packages.

vkallesh replied:
Done. Restore to previous url value.

dev-zero commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -26,41 +26,151 @@ class Amdfftw(FftwBase):
+    conflicts(
+        '+amd-app-opt',
+        when='precision=quad',
+        msg='amd-trans is not supported for quad precision')
```
That whole thing could be simplified. Ex.

```
    when('+amd-top-n-planner'):
        conflicts('+threads', msg='amd-top-n-planner does not support threads')
        ...
```

But maybe someone else has even better ideas? Since the `msg` are mostly repetitive you could also simply loop over the conflict condition strings.

vkallesh replied:
This change would call for major code refactoring. Need to revalidate again for all scenarios.

Can I takeup this change in coming release ?

During next release, would follow, following reference spack recipe (which has above suggested change):
https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/quantum-espresso/package.py

tldahlgren replied:
I agree with @dev-zero about simplifying this but am also fine with the change being deferred to another PR.

dev-zero commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -106,18 +216,43 @@ def configure(self, spec, prefix):
+            options.append('--enable-amd-app-opt')
+        else:
+            options.append('--disable-amd-app-opt')
+
```
Use `enable_or_disable` instead? https://spack.readthedocs.io/en/latest/build_systems/autotoolspackage.html#enable-or-disable

vkallesh replied:
Done

amd-toolchain-support committed:
Incorporated review comments

1) Restore to previous url value
2) Instead of if else statements, used spack's enable_or_disable feature

vkallesh commented:
I would like others to review this pull request as I worked on updating spack recipes for AOCL 3.1 version.

tldahlgren commented on var/spack/repos/builtin/packages/amdblis/package.py:
```diff
@@ -22,17 +22,50 @@ class Amdblis(BlisBase):
+
+        """ To enabled Fortran to C calling convention for
+        complex types when compiling with aocc flang"""
+        if "@3.0 %aocc" in self.spec:
```
We generally recommend using `satisfies()` for spec checks that are more than whether a variant is enabled:

```suggestion
        if self.spec.satisfies("@3.0 %aocc"):
```

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/amdblis/package.py:
```diff
@@ -22,17 +22,50 @@ class Amdblis(BlisBase):
+        if "@3.0 %aocc" in self.spec:
             args.append('CFLAGS={0}'.format("-DAOCL_F2C"))
             args.append('CXXFLAGS={0}'.format("-DAOCL_F2C"))
+        elif "@3.0.1: %aocc" in self.spec:
```
```suggestion
        elif self.spec.satisfies("@3.0.1: %aocc"):
```

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/amdblis/package.py:
```diff
@@ -22,17 +22,50 @@ class Amdblis(BlisBase):
+        elif "@3.0.1: %aocc" in self.spec:
+            args.append('--complex-return=intel')
+
+        if "@3.1: " in self.spec:
```
```suggestion
        if self.spec.satisfies("@3.1:"):
```

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/amdblis/package.py:
```diff
@@ -22,17 +22,50 @@ class Amdblis(BlisBase):
+
+        # "amdzen" - A fat binary or multiarchitecture binary
+        # support for 3.1 release onwards
+        if "@3.1: " in self.spec:
```
```suggestion
        if self.spec.satisfies("@3.1:"):
```

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -31,36 +31,146 @@ class Amdfftw(FftwBase):
+    conflicts(
+        'precision=quad',
+        when='@2.2 %aocc',
+        msg='AOCC clang version 2.2 is not supported for quad precision')
```
This is a nit for all of the messages phrased as "not supported for".  Is it not the case where the option or mode is not supported by the version(s) or variant(s)?  Examples:

- `msg='Quad precision is not supported by AOCC clang version 2.2'`; and
- `msg='long_double precision is not supported with amd-fast-planner'`.

tldahlgren commented on var/spack/repos/builtin/packages/amdlibflame/package.py:
```diff
@@ -64,8 +75,29 @@ def configure_args(self):
+
         return args
 
+    @run_after('build')
+    @on_package_attributes(run_tests=True)
```
These directives should not be needed.  That is, it should be sufficient to override the `check` method.

vkallesh replied:
This step was added to ensure 'make check' command invokes after build and before install command.

tldahlgren replied:
I realize that.  My point was that the packages you're inheriting from already have the `check` method set up to run after the `build` phase and for the `--tests`(or deprecated `--run-tests`) options.  

There is no harm in including them here, which reminds me that I have noticed some packages run their checks twice.  Perhaps this is why.  (I'll have to look into that when I get some spare cycles.)

tldahlgren commented on var/spack/repos/builtin/packages/amdlibm/package.py:
```diff
@@ -25,39 +25,45 @@ class Amdlibm(SConsPackage):
     patch("0002-libm-ose-prevent-log-v3.c-from-building.patch", when="@2.2")
 
-    conflicts("%gcc@:9.1", msg="Minimum required GCC version is 9.2.0")
+    conflicts("%gcc@9.1.0:11.2.0", msg="Supported GCC versions are from 9.2.0 to 11.1.0")
```
This is a `conflict` so shouldn't the original be retained to catch any attempts to use a `gcc` below the minimum and add a new conflict to ensure any attempts to use a `gcc` above the maximum are flagged?

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/amdlibm/package.py:
```diff
@@ -25,39 +25,45 @@ class Amdlibm(SConsPackage):
+        if "@:3.0 %aocc" in spec:
+            args.append("--compiler=aocc")
+
+        if "@:3.0 " in spec:
```
Still recommend use of `spec.satisfies()` for version ranges.

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/amdlibm/package.py:
```diff
@@ -25,39 +25,45 @@ class Amdlibm(SConsPackage):
+            args.append("CXX={0}".format(self.compiler.cxx))
+        else:
+            args.append("ALM_CC={0}".format(self.compiler.cc))
+            args.append("ALM_CXX={0}".format(self.compiler.cxx))
```
Another nit.  Could simplify e.g.:

```suggestion
        var_prefix = '' if spec.satisfies("@:3.0 ") else 'ALM_'
        args.append("{0}CC={1}".format(var_prefix, self.compiler.cc))
        args.append("0}CXX={1}".format(var_prefix, self.compiler.cxx))
```

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/amdscalapack/package.py:
```diff
@@ -47,8 +62,13 @@ def cmake_args(self):
             args.extend(['-DUSE_DOTC_WRAPPER:BOOL=%s' % (
                         'ON' if spec.satisfies('%aocc ^amdblis') else 'OFF')])
 
+        if "+ilp64" in spec:
+            args.extend(['-DENABLE_ILP64=ON'])
+
+        if "@:3.0" in spec:
+            args.extend(['-DUSE_F2C=ON'])
```
You could use the `CMakePackage` helpers (https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html#adding-flags-to-cmake):

```suggestion
        args.extend([self.define_from_variant('ENABLE_ILP64', 'ilp64')])  # -DENABLE_ILP64:BOOL=ON
        args.extend([self.define('USE_F2C', spec.satisfies('@:3.0'))])  # -DUSE_F2C:BOOL=ON
```

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/amdscalapack/package.py:
```diff
@@ -20,19 +20,34 @@ class Amdscalapack(ScalapackBase):
     version('3.0', sha256='6e6f3578f44a8e64518d276e7580530599ecfa8729f568303ed2590688e7096f')
     version('2.2', sha256='2d64926864fc6d12157b86e3f88eb1a5205e7fc157bf67e7577d0f18b9a7484c')
 
+    def url_for_version(self, version):
+        if version == Version('3.1'):
+            return "https://github.com/amd/aocl-scalapack/archive/3.1.tar.gz"
+        elif version == Version('3.0'):
+            return "https://github.com/amd/scalapack/archive/3.0.tar.gz"
+        elif version == Version('2.2'):
+            return "https://github.com/amd/scalapack/archive/2.2.tar.gz"
+
```
It is preferable to have the directives listed *before* methods so can this be moved to appear after the `conflicts` directive so it is included in the methods "section"?

vkallesh replied:
Incorporated review comments

tldahlgren reviewed (changes requested):
I confirmed all of the sha256 are correct.  I do have several suggestions, some of which are more nits.

Let me know if you have any questions or concerns with any of them.

amd-toolchain-support committed:
Incorporated following review comments:

1. Use of satisfies() for spec checks
2. Seperate conflict statements to check for minimum and maximum GCC versions
3. Used CMakePackage helpers
4. Code rearrangement to have the directives listed before methods

tldahlgren reviewed (approved):
LGTM. Thanks.
